Content.Server/Singularity/EntitySystems/SingularityGeneratorSystem.cs

```diff
@@ -113,10 +149,51 @@ private void HandleParticleCollide(EntityUid uid, ParticleProjectileComponent co
                     ParticleAcceleratorPowerState.Level3 => 8,
                     _ => 0
                 },
-                singularityGeneratorComponent
+                generatorComp
             );
-            EntityManager.QueueDeleteEntity(uid);
-        }
+        EntityManager.QueueDeleteEntity(uid);
+    }
+
+    private void OnEmagged(EntityUid uid, SingularityGeneratorComponent component, ref GotEmaggedEvent args)
+    {
+        _popupSystem.PopupEntity(Loc.GetString("comp-generator-failsafe-disabled", ("target", uid)), uid);
+        component.FailsafeDisabled = true;
+        args.Handled = true;
     }
     #endregion Event Handlers
+
+    /// <summary>
+    /// Checks whether there's a containment field in a given direction away from the generator
+    /// </summary>
+    /// <param name="transform">The transform component of the singularity generator.</param>
+    /// <remarks>Mostly copied from <see cref="ContainmentFieldGeneratorSystem"/> </remarks>
+    private bool CheckContainmentField(Direction dir, Entity<SingularityGeneratorComponent> generator, TransformComponent transform)
+    {
+        var component = generator.Comp;
+
+        var (worldPosition, worldRotation) = _transformSystem.GetWorldPositionRotation(transform);
+        var dirRad = dir.ToAngle() + worldRotation;
+
+        var ray = new CollisionRay(worldPosition, dirRad.ToVec(), component.CollisionMask);
+        var rayCastResults = _physics.IntersectRay(transform.MapID, ray, component.FailsafeDistance, generator, false);
+        var genQuery = GetEntityQuery<ContainmentFieldComponent>();
+
+        RayCastResults? closestResult = null;
+
+        foreach (var result in rayCastResults)
+        {
+            if (genQuery.HasComponent(result.HitEntity))
+                closestResult = result;
+
+            break;
+        }
+
+        if (closestResult == null)
+            return false;
+
+        var ent = closestResult.Value.HitEntity;
+
+        // Check that the field can't be moved. The fields' transform parenting is weird, so skip that
+        return TryComp<PhysicsComponent>(ent, out var collidableComponent) && collidableComponent.BodyType == BodyType.Static;
+    }
```

Comment:
This is a bit skrunkly but probably fine for now in lieu of a better raycast API.

Reply:
Yeah this is basically reusing the `ContainmentFieldGeneratorSystem` stuff, which I at some point looked at before and was familiar with x3

Not sure if raycast or area check would have been better, but yeah, this would require engineers to make sure all fields are up.